### DISCUSSION TITLE
Referencing Multiple DBT Profiles in dbt_assets

### DISCUSSION THREAD

**Discussion by fredonia88** [dagster-io/dagster#19045] [Q&A] [answered]
What is the best way to to create a single Dagster schedule or job that will materialize dbt assets that use two different profiles?

Assume my dbt project contains assets in three directories: dir_a, dir_b, and dir_c. dir_b contains dbt assets that should be materialized with profile_b. All remaining directories should use the default target (profile_a). However, dir_b assets are dependent on upstream assets materialized in dir_a.

I'm defining the assets, like this:
```
@dbt_assets(manifest=dbt_manifest_path)
def my_dbt_assets(context: OpExecutionContext, dbt: DbtCliResource):
    yield from dbt.cli(['build', '--exclude=dir_b', '--indirect-selection=buildable'], context=context).stream()
    yield from dbt.cli(['build', '--select=dir_b', '--indirect-selection=buildable', '--target=profile_b'], context=context).stream()
```

I'm creating a schedule to materialize my_dbt_assets, like this:
```
daily_dbt_assets_schedule = build_schedule_from_dbt_selection(
    [my_dbt_assets],
    job_name='daily_dbt_build',
    cron_schedule='15 5 * * *',
    dbt_select='tag:daily'
)
```

I do want to limit this schedule to only assets that should materialize daily. However, this schedule doesn't work because the`dbt_select='tag:daily'` is adding an extra select to the end of the second dbt build command, like so:

```
dbt build --select=dir_b --indirect-selection=buildable --target=profile_b --select fqn:some_asset_from_dir_a,some_asset_from_dir_c,etc
```

I tried creating two separate jobs for dir_a and dir_c, and dir_b. However, I don't think this will work, cause this breaks the dependency between dir_b and dir_a.

How do others handle multiple profiles in a single dbt project? I am definitely open to different suggestions or approaches, and am happy to provide more information and context.

Thanks!

**Comment by rexledesma** [comment, marked as answer]
Try splitting up your `my_dbt_assets` definition by your selection. Then, you can schedule your assets by tag.

```python
@dbt_assets(manifest=dbt_manifest_path, select="dir_b")
def dir_b_dbt_assets(context: OpExecutionContext, dbt: DbtCliResource):
    yield from dbt.cli(['build', '--indirect-selection=buildable', '--target=profile_b'], context=context).stream()
    
@dbt_assets(manifest=dbt_manifest_path, exclude="dir_b")
def my_dbt_assets(context: OpExecutionContext, dbt: DbtCliResource):
    yield from dbt.cli(['build', '--indirect-selection=buildable'], context=context).stream()
    
    
daily_dbt_assets_schedule = build_schedule_from_dbt_selection(
    [my_dbt_assets, dir_b_dbt_assets],
    job_name='daily_dbt_build',
    cron_schedule='15 5 * * *',
    dbt_select='tag:daily'
)
```

**Reply by fredonia88** [reply]
Thank you @rexledesma for the response. I tried implementing that, but am getting an error that complains `build_schedule_from_dbt_selection` has duplicate asset definitions: `check.invariant(len(dbt_assets) == 1, "Exactly one dbt AssetsDefinition is required")`

I also tried selecting a single asset in each asset definition, but still get the above error (asset_b and asset_a names are definitely unique): 

```
@dbt_assets(manifest=dbt_manifest_path, select="asset_b")
def dir_b_dbt_assets(context: OpExecutionContext, dbt: DbtCliResource):
    yield from dbt.cli(['build', '--indirect-selection=buildable', '--target=profile_b'], context=context).stream()
    
@dbt_assets(manifest=dbt_manifest_path, select="asset_a")
def my_dbt_assets(context: OpExecutionContext, dbt: DbtCliResource):
    yield from dbt.cli(['build', '--indirect-selection=buildable'], context=context).stream()
```

**Reply by rexledesma** [reply]
Could you just construct the schedule using the the `ScheduleDefinition` and `DbtManifestAssetSelection`?

```python
daily_dbt_assets_schedule = ScheduleDefinition(
    cron_schedule="15 5 * * *",
    job=define_asset_job(
        name="daily_dbt_build",
        selection=DbtManifestAssetSelection(
            manifest=dbt_manifest_path,
            select="tag:daily",
        ),
    ),
)
```

**Reply by fredonia88** [reply]
Thank you @rexledesma! That worked.

I have dbt assets in `dir_b_dbt_assets` that are dependent on dbt assets in `my_dbt_assets`, but I cannot use dbt's `ref()` function. This is because I'm using a different profile for `dir_b_dbt_assets`, and that profile's target schema is different. 

I have defined my_dbt_assets as a source for dir_b_dbt_assets, and am thinking I can use a dbt freshness policy to make them dependent. But is there a way to explicitly make `dir_b_dbt_assets` dependent on `my_dbt_assets` in Dagster either at the asset or job level (if I break these up into two jobs)? Happy to start a new thread if you prefer.

**Reply by rexledesma** [reply]
Have you tried forcing dependencies in dbt? This way, you can continue visualizing the dependencies at the asset level. See https://docs.getdbt.com/reference/dbt-jinja-functions/ref#forcing-dependencies for an example.

**Reply by fredonia88** [reply]
Thanks, yet again @rexledesma. That worked.
